File: Java/Threaded/src/sorting_algorithms/QuickSort.java
package sorting_algorithms;

/**
 * Created by pauljohanneskraft on 28.02.16.
 */
public class QuickSort<T extends Comparable> extends Sort<T> {

    public QuickSort(T[] array) {
        this(array, 0, array.length - 1);
    }

    private QuickSort(T[] array, int left, int right) {
        super(array);
        this.left = left;
        this.right = right;
    }

    public void setArray(T[] array) {
        super.setArray(array);
        this.left = 0;
        this.right = length - 1;
    }

    public void run() {
        if(left < right) {
            int pivot = partition();
            int right = this.right;
            this.right = pivot - 1;
            run();
            this.left = pivot + 1;
            this.right = right;
            run();
        }
    }

    private int partition() {
        int i = left;
        int j = right - 1;
        T p = array[right];
        do {
            while(array[i].compareTo(p) <= 0 && i < right) { i++; }
            while(array[j].compareTo(p) >= 0 && j > left) { j--; }
            if(i < j) { swap(i,j); }
        } while(i < j);
        if(array[i].compareTo(p) > 0) { swap(i, right); }
        return i;
    }

    public String toString() { return "QuickSort"; }
}

File: C++/sorting algorithms/QuickSort.cpp
//
// Created by Paul Kraft on 16.03.16.
//

#include "QuickSort.h"

void QuickSort::sort(Comparable* array[]) {
    sortPart(array, 0, (sizeof(array)/sizeof(array[0])) - 1);
}

void QuickSort::sortPart(Comparable* array[], int left, int right) {
    if(left < right) {
        int pivot = partition(array, left, right);
        sortPart(array, left, pivot - 1);
        sortPart(array, pivot + 1, right);
    }
}

int QuickSort::partition(Comparable* array[], int left, int right) {
    int i = left;
    int j = right;
    Comparable* tmp;
    Comparable* pivot = array[(left + right) / 2];

    /* partition */
    while (i <= j) {
        while (array[i] < pivot)
            i++;
        while (array[j] > pivot)
            j--;
        if (i <= j) {
            tmp = array[i];
            array[i] = array[j];
            array[j] = tmp;
            i++;
            j--;
        }
    };
    return i;
}